openedx/core/djangoapps/enrollments/views.py

```diff
@@ -987,3 +994,164 @@ def get_queryset(self):
+
+            If there's an error while getting the submission history an empty response will
+            be returned.
+            The submission history response has the following attributes:
+
+                * Results: A list of submission history:
+                    * course_id: Course id
+                    * course_name: Course name
+                    * user: Username
+                    * problems: List of problems
+                        * location: problem location
+                        * name: problem's display name
+                        * submission_history: List of submission history
+                            * state: State of submission.
+                            * grade: Grade.
+                            * max_grade: Maximum possible grade.
+                        * data: problem's data.
+        """
+        username = request.GET.get('username', request.user.username)
+        data = []
+        if GlobalStaff().has_user(request.user):
+            all_users = bool(request.GET.get('all', False))
+        else:
+            all_users = False
+        course_id = request.GET.get('course_id')
+
+        if not (all_users or username == request.user.username or GlobalStaff().has_user(request.user) or
+                self.has_api_key_permissions(request)):
+            return Response(data)
+
+        course_enrollments = CourseEnrollment.objects.select_related('user').filter(is_active=True)
+        if course_id:
+            if not course_id.startswith("course-v1:"):
+                course_id = "course-v1:{}".format(course_id)
+            try:
+                course_enrollments = course_enrollments.filter(
+                    course_id=CourseLocator.from_string(course_id.replace(' ', '+'))
+                ).order_by('created')
+            except KeyError:
+                return Response(data)
+
+        if not all_users:
+            course_enrollments = course_enrollments.filter(user__username=username).order_by('created')
+
+        courses = {}
+        for course_enrollment in course_enrollments:
+            try:
+                course_list = courses.get(course_enrollment.course_id)
+                if course_list:
+                    course, course_children = course_list
+                else:
+                    course = get_course(course_enrollment.course_id, depth=4)
+                    course_children = course.get_children()
+                    courses[course_enrollment.course_id] = [course, course_children]
+            except ValueError:
+                continue
+            course_data = self._get_course_data(course_enrollment, course, course_children)
+            data.append(course_data)
+
+        return Response({'results': data})
+
+    def _get_problem_data(self, course_enrollment, component):
+        """
+        Get problem data from a course enrollment.
+
+        Args:
+        -----
+        course_enrollment: Course Enrollment.
+        component: Component to analyze.
```

Comment:
```suggestion
        block: XBlock to analyze.
```

Let's keep the naming unified.